custom_components/elasticsearch/es_index_manager.py

```diff
@@ -30,35 +39,100 @@ def __init__(self, hass, config, gateway):
             return
-        self._using_ilm = self._config.get(CONF_ILM_ENABLED)
 
-        await self._create_index_template()
+        if self.index_mode == INDEX_MODE_LEGACY:
+            self._using_ilm = self._config.get(CONF_ILM_ENABLED)
+
+            await self._create_legacy_template()
 
-        if self._using_ilm:
-            await self._create_ilm_policy(self._config)
+            if self._using_ilm:
+                await self._create_ilm_policy(self._config)
+        else:
+            await self._create_index_template()
 
         LOGGER.debug("Index Manager initialized")
 
     async def _create_index_template(self):
         """Initialize the Elasticsearch cluster with an index template, initial index, and alias."""
         from elasticsearch7.exceptions import ElasticsearchException
 
+        LOGGER.debug("Initializing modern index templates")
+
+        if not self._gateway.es_version.meets_minimum_version(major=8, minor=7):
+            raise ElasticException(
+                "A version of Elasticsearch that is not compatible with TSDS datastreams detected (%s). Use Legacy Index mode.",
+                f"{self._gateway.es_version.major}.{self._gateway.es_version.minor}",
+            )
+
+        client = self._gateway.get_client()
+
+        # Open datastreams/index_template.json and load the ES modern index template
+        with open(
+            os.path.join(
+                os.path.dirname(__file__), "datastreams", "index_template.json"
+            ),
+            encoding="utf-8",
+        ) as json_file:
+            index_template = json.load(json_file)
+
+        # Check if the index template already exists
+        existingTemplate = await client.indices.get_index_template(
+            name=DATASTREAM_METRICS_INDEX_TEMPLATE_NAME, ignore=[404]
+        )
+        LOGGER.debug("got template response: " + str(existingTemplate))
+
+        if existingTemplate:
+            LOGGER.debug("Updating index template")
+        else:
+            LOGGER.debug("Creating index template")
+
+        try:
+            await client.indices.put_index_template(
+                name=DATASTREAM_METRICS_INDEX_TEMPLATE_NAME, body=index_template
+            )
+
+        except ElasticsearchException as err:
+            LOGGER.exception("Error creating/updating index template: %s", err)
+            # We do not want to proceed with indexing if we don't have any index templates as this
+            # will result in the user having to clean-up indices with improper mappings.
+            if not existingTemplate:
+                raise err
+
+    async def _create_legacy_template(self):
+        """Initialize the Elasticsearch cluster with an index template, initial index, and alias."""
+        from elasticsearch7.exceptions import ElasticsearchException
+
+        LOGGER.debug("Initializing legacy index templates")
+
+        if self._gateway.es_version.is_serverless():
+            raise ElasticsearchException(
```

Comment:
Let's not raise an exception owned by a dependency. We can raise `ElasticException` from `custom_components.elasticsearch.errors`

```suggestion
            raise ElasticException(
```

Reply:
was in the middle of doing this as i was updating the tests, will be included shortly